server/src/main/java/org/apache/uniffle/server/ShuffleServer.java

```diff
@@ -386,21 +387,21 @@ private void waitDecommissionFinish() {
   }
 
   public synchronized void cancelDecommission() {
-    if (!isDecommissioning()) {
+    boolean wasDecomissioning = serverStatus.compareAndSet(ServerStatus.DECOMMISSIONING, ServerStatus.ACTIVE);
+    boolean wasDecomissioned = serverStatus.compareAndSet(ServerStatus.DECOMMISSIONED, ServerStatus.ACTIVE);
+    if (!wasDecomissioning && !wasDecomissioned) {
       LOG.info("Shuffle server is not decommissioning. Nothing needs to be done.");
       return;
     }
-    if (ServerStatus.DECOMMISSIONED.equals(serverStatus.get())) {
-      serverStatus.set(ServerStatus.ACTIVE);
-      return;
-    }
-    serverStatus.set(ServerStatus.ACTIVE);
```

Comment:
State could change between `isDecommissioning()` and `serverStatus.set(ServerStatus.ACTIVE)`.